Add support for custom provider hook (#82)
matheusml/zsh-ai: INSTALL.md

`ZSH_AI_OPENAI_REASONING_EFFORT` sets the `reasoning_effort` parameter for OpenAI-compatible endpoints that support it. For example, Groq supports `none` and `default` for `qwen/qwen3.6-27b`; set `none` to disable reasoning tokens. Common values are `none`, `default`, `minimal`, `low`, `medium`, and `high`; leave unset for upstream defaults.

## Custom Provider

Set `ZSH_AI_PROVIDER` to `custom` and define `_zsh_ai_query_custom` before
`zsh-ai` loads. The function receives the natural-language query as its first
argument. It must print only the generated command to standard output, and

> return a non-zero status when the provider fails, optionally printing an error message to standard output.

Reviewer: the plugin never sees the exit status: `_zsh_ai_execute_command` does `local cmd=$(_zsh_ai_query "$query")`, which swallows it, and failure is detected purely by prefix (`"$cmd" != "Error:"*` in utils.zsh and widget.zsh). so a function that follows this doc and echoes a failure message without the `Error:` prefix gets that message inserted into the buffer as a runnable command.

simplest fix is documenting the prefix convention:

```suggestion
argument. It must print only the generated command to standard output. If the
provider fails, it must print a message starting with `Error:`; the plugin
detects failure by that prefix and ignores the function's exit status.
```

the nicer fix is making `_zsh_ai_execute_command` respect the exit status, but that touches every provider. fine with either, which were you thinking?

Author: I think respecting the exit status is a cleaner fix. The error checks in `utils.zsh` and `widget.zsh` are either-or checks for non-zero status or `Error:`-prefixed output, so I'll present both options here. Also, `_zsh_ai_execute_command` can be removed as an intermediary, since it only seems to set an exit code.

This example uses the Codex CLI:

```zsh
export ZSH_AI_PROVIDER="custom"

_zsh_ai_query_custom() {
	local query="${1-}"
	local system_prompt_file response

	local system_prompt_file=$(mktemp)

> 	_zsh_ai_get_system_prompt > "$system_prompt_file"

Reviewer: did you mean to skip the context? `_zsh_ai_get_system_prompt` takes it as `$1`, so this builds a prompt with an empty `Context:` section, and codex never sees the cwd/git/os info the built-in providers get.

```suggestion
	_zsh_ai_get_system_prompt "$(_zsh_ai_build_context)" > "$system_prompt_file"
```

if skipping it was deliberate, worth a line saying so.

Author: Didn't use context because Codex automatically adds its own context into the prompt such as the cwd and shell name, but on taking a closer look, the Codex context does not overlap the information provided `_zsh_ai_build_context`, so I'll incorporate your suggestion.

	response=$(
		{
			codex exec \
				--ephemeral \
				--ignore-user-config \
				--skip-git-repo-check \
				--sandbox read-only \
				--model 'gpt-5.6-luna' \
				--config model_reasoning_effort=low \
				--config "model_instructions_file=\"$system_prompt_file\"" \

> 				-- "$query" \

Author: I'll need to modify this to ensure that Codex does not read an untrusted 'AGENTS.md' file.

				2>/dev/null
		} always {
			rm -f -- "$system_prompt_file"
		}
	)

	if (( status != 0 )); then
		echo "Error: Something went wrong while running codex."
		return "$status"

> 	fi

Reviewer: this returns 0 on failure: the `echo` resets `$status` before the `return` reads it.

```suggestion
	local ret=$status
	if (( ret != 0 )); then
		echo "Error: Something went wrong while running codex."
		return "$ret"
	fi
```

Author: Good catch! Will fix.
